### PR TITLE
[pinot docker image] Upgrade prometheus and scope rulesets to components

### PR DESCRIPTION
**Pull request by richardstartin**
We frequently see that prometheus is a resource hog in our customer deployments, with regular expression matching leading to excessive allocation and CPU time. Pinot currently uses version 0.12.0, which is hardcoded into deployments and the jar is checked in. 
Screenshot 2022-02-17 at 14 31 09

Screenshot 2022-02-18 at 08 43 34

Prometheus fixed this here, allowing : https://github.com/prometheus/jmx_exporter/pull/518

This PR changes this so that the latest version (pinned at 0.16.1) of jmx_prometheus_javaagent is downloaded in the docker image, and caching is enabled for our rules.

### CODE REVIEW

**Comment by richardstartin** [comment]
> btw we shouldn't check in a binary jar to repo in the first place :-)

you're telling me?

**Comment by richardstartin** [comment]
Adding the cache expressions needs thorough testing as it is not appropriate to all metric matching logic, I suggest delaying caching

**Comment by xiangfu0** [comment]
Makes a lot of sense! Thanks Richard!

File: docker/images/pinot/Dockerfile [hunk 1]
```diff
@@ -79,6 +79,9 @@ COPY bin ${PINOT_HOME}/bin
 COPY etc ${PINOT_HOME}/etc
 COPY examples ${PINOT_HOME}/examples
 
+RUN wget -O ${PINOT_HOME}/etc/jmx_prometheus_javaagent-0.16.1.jar https://repo1.maven.org/maven2/io/prometheus/jmx/jmx_prometheus_javaagent/0.16.1/jmx_prometheus_javaagent-0.16.1.jar && \
```
Comment (xiangfu0) [inline docker/images/pinot/Dockerfile:82]: For backward compatibility, can we also download javaagent-0.12.0 as well?
So users on the latest image won't fail suddenly during restart.

**Comment by xiangfu0** [comment]
> > btw we shouldn't check in a binary jar to repo in the first place :-)
> 
> you're telling me?

This is my bad :p

File: docker/images/pinot/Dockerfile (hunk 1 above)
Comment (richardstartin) [reply docker/images/pinot/Dockerfile:82]: Great idea

**Review by xiangfu0 — APPROVED** [review]

**Comment by richardstartin** [comment]
I figured out that the much bigger problem here is that all the rules are evaluated on all the instance types, so we are running controller rules on the brokers, controller + broker on the servers, controller + broker + server rules on the minions before matching. We should break this down before considering caching.

File: docker/images/pinot/Dockerfile (hunk 1 above)
Comment (richardstartin) [reply docker/images/pinot/Dockerfile:83]: done

File: docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+- pattern: "\"org.apache.pinot.common.metrics\"<type=\"ControllerMetrics\", name=\"pinot.controller.tableStorageQuotaUtilization.(\\w+)_(\\w+)\"><>(\\w+)"
+  name: "pinot_controller_tableStorageQuotaUtilization_$3"
+  labels:
+    table: "$1"
+    tableType: "$2"
+- pattern: "\"org.apache.pinot.common.metrics\"<type=\"ControllerMetrics\", name=\"pinot.controller.tableStorageEstMissingSegmentPercent.(\\w+)_(\\w+)\"><>(\\w+)"
+  name: "pinot_controller_tableStorageEstMissingSegmentPercent_$3"
+  labels:
+    table: "$1"
+    tableType: "$2"
```
Comment (xiangfu0) [inline docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml:100]: nit: newline

File: docker/images/pinot/etc/jmx_prometheus_javaagent/configs/broker.yml [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    table: "$1"
+    tableType: "$2"
+    partition: "$3"
+
+- pattern: "\"org.apache.pinot.common.metrics\"<type=\"MinionMetrics\", name=\"pinot.minion.(\\w+)\"><>(\\w+)"
+  name: "pinot_minion_$1_$2"
+- pattern: "\"org.apache.pinot.common.metrics\"<type=\"MinionMetrics\", name=\"pinot.minion.(\\w+).(\\w+)\"><>(\\w+)"
+  name: "pinot_minion_$2_$3"
+  labels:
+    id: "$1"
```
Comment (xiangfu0) [inline docker/images/pinot/etc/jmx_prometheus_javaagent/configs/broker.yml:171]: nit: newline

**Pull request merged**